File: src/omnia/data/structure/TypedMap.java
package omnia.data.structure;

import static java.util.Objects.requireNonNull;

import java.util.Optional;
import omnia.data.structure.immutable.ImmutableMap;
import omnia.data.structure.mutable.HashMap;
import omnia.data.structure.mutable.MutableMap;

public final class TypedMap {

  public static final class Key<V> {

    public static <V> Key<V> create() {
      return new Key<>();
    }
  }

  public static class Builder {
    private final MutableMap<Key<?>, Object> map = HashMap.create();

    public <V> Builder putMapping(Key<V> key, V value) {
      requireNonNull(key);
      requireNonNull(value);
      map.putMapping(key, value);
      return this;
    }

    public Builder putMappings(TypedMap map) {
      map.map.entries().forEach(entry -> this.map.putMapping(entry.key(), entry.value()));
      return this;
    }

    public Builder remove(Key<?> key) {
      requireNonNull(key);
      map.removeKey(key);
      return this;
    }

    public TypedMap build() {
      return new TypedMap(map);
    }
  }

  public static Builder builder() {
    return new Builder();
  }

  public static TypedMap of() {
    return new TypedMap(ImmutableMap.empty());
  }

  public static <V> TypedMap of(Key<V> key, V value) {
    return new TypedMap(ImmutableMap.of(key, value));
  }

  private final Map<Key<?>, Object> map;

  private TypedMap(Map<Key<?>, Object> map) {
    this.map = ImmutableMap.copyOf(map);
  }

  public <V> Optional<V> get(Key<V> key) {
    @SuppressWarnings("unchecked")
    Optional<V> value = map.valueOf(key).map(v -> (V) v);
    return value;
  }

  public Builder toBuilder() {
    return new Builder().putMappings(this);
  }
}

File: src/omnia/data/structure/TypedMap.kt
package omnia.data.structure

import omnia.data.structure.immutable.ImmutableMap
import omnia.data.structure.mutable.HashMap
import omnia.data.structure.mutable.MutableMap
import java.util.Optional
import java.util.function.Consumer

class TypedMap private constructor(map: Map<Key<*>, Any?>) {
    class Key<V> {
        fun <V> create(): Key<V> {
            return Key()
        }
    }

    class Builder {
        private val map: MutableMap<Key<*>, Any?> = HashMap.create()
        fun <V> putMapping(key: Key<V>, value: V): Builder {
            map.putMapping(key, value)
            return this
        }

        fun putMappings(map: TypedMap): Builder {
            map.map.entries().forEach(Consumer { entry -> this.map.putMapping(entry.key(), entry.value()) })
            return this
        }

        fun remove(key: Key<*>): Builder {
            map.removeKey(key)
            return this
        }

        fun build(): TypedMap {
            return TypedMap(map)
        }
    }

    private val map: Map<Key<*>, Any?>
    operator fun <V> get(key: Key<V>): Optional<V> {
        return map.valueOf(key).map { v ->
            @Suppress("UNCHECKED_CAST")
            v as V
        }
    }

    fun toBuilder(): Builder {
        return Builder().putMappings(this)
    }

    companion object {
        fun builder(): Builder {
            return Builder()
        }

        fun of(): TypedMap {
            return TypedMap(ImmutableMap.empty())
        }

        fun <V> of(key: Key<V>, value: V): TypedMap {
            return TypedMap(ImmutableMap.of(key, value))
        }
    }

    init {
        this.map = ImmutableMap.copyOf(map)
    }
}